### PR TITLE
Feature/selective sync

### PR DESCRIPTION
**Pull request by cshfang**
*Issue #:* #298

*Description of changes:*
Currently, DataStore downloads the entire contents of your cloud data source to local device by default when using Amplify Flutter. In comparison, customers of native Amplify Android and iOS SDKs are able to configure DataStore to selectively persist only a subset of data instead.

This PR adds selective sync to Amplify Flutter. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt [hunk 1]
```diff
@@ -70,23 +72,32 @@ class AmplifyDataStorePlugin : FlutterPlugin, MethodCallHandler {
     }
 
     @VisibleForTesting
-    constructor(eventHandler: DataStoreObserveEventStreamHandler,
-            hubEventHandler: DataStoreHubEventStreamHandler) {
+    constructor(
```
[the remark this answers is not in the document]
Comment (cshfang) [reply packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt:75]: Hmm, I didn't install any additional linters/formatters. Definitely realize the issue with formatting changes making things difficult to review... in the last iteration of this PR, I tried to call out the sections of code that were relevant (i.e. non-formatting-only) and these files were non-dart so I didn't call them out this time around as they shouldn't have changed. But you should be able to focus on just the `onConfigureDataStore` and `buildSyncExpressions` functions (and their equivalents in swift).

File: packages/amplify_datastore/lib/method_channel_datastore.dart [hunk 2]
```diff
@@ -32,21 +32,42 @@ class AmplifyDataStoreMethodChannel extends AmplifyDataStore {
...
+  Future<void> configureDataStore({
+    ModelProviderInterface? modelProvider,
+    List<DataStoreSyncExpression>? syncExpressions,
+    int? syncInterval,
+    int? syncMaxRecords,
+    int? syncPageSize,
+  }) async {
+    _channel.setMethodCallHandler((MethodCall call) async {
+      switch (call.method) {
+        case 'resolveQueryPredicate':
```
[the remark this answers is not in the document]
Comment (cshfang) [reply packages/amplify_datastore/lib/method_channel_datastore.dart:44]: Sync expressions need to be able to be re-evaluated at runtime ([iOS](https://docs.amplify.aws/lib/datastore/sync/q/platform/ios#reevaluate-expressions-at-runtime), [Android](https://docs.amplify.aws/lib/datastore/sync/q/platform/android#reevaluate-expressions-at-runtime)). So when the DataStore is restarted and sync expressions are re-evaluated, we needed a way to call back to Flutter and say, re-run those predicate resolvers and give me back the new predicates.

**Review by HuiSF — APPROVED** [review]

File: packages/amplify_datastore/lib/method_channel_datastore.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jordan-Nelson) [reply packages/amplify_datastore/lib/method_channel_datastore.dart:44]: I see. I did not know they needed to be re-evaluated at runtime.

**Review by Jordan-Nelson — APPROVED** [review]

**Review by ragingsquirrel3 — APPROVED** [review]

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt [hunk 3]
```diff
@@ -118,31 +131,39 @@ class AmplifyDataStorePlugin : FlutterPlugin, MethodCallHandler {
...
+                                " are not available. Request: " + request.toString()
+                    )
+                )
             }
             return
         }
 
         val modelSchemas: List<Map<String, Any>> = request["modelSchemas"].safeCastToList()!!
+        val syncExpressions: List<Map<String, Any>> =
+            request["syncExpressions"].safeCastToList() ?: emptyList()
```
Comment (ragingsquirrel3) [inline packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt:153]: nice, thanks!

**Pull request merged**